src/openrct2/interface/Window.cpp

```diff
@@ -205,76 +210,40 @@ void WindowSetWindowLimit(int32_t value)
  */
 void WindowClose(WindowBase& w)
 {
-    auto itWindow = WindowGetIterator(&w);
-    if (itWindow == g_window_list.end())
-        return;
-
-    // Explicit copy of the shared ptr to keep the memory valid.
-    std::shared_ptr<WindowBase> window = *itWindow;
-
-    WindowEventCloseCall(window.get());
+    WindowEventCloseCall(&w);
 
     // Remove viewport
-    window->RemoveViewport();
+    w.RemoveViewport();
 
     // Invalidate the window (area)
-    window->Invalidate();
+    w.Invalidate();
 
-    // The window list may have been modified in the close event
-    itWindow = WindowGetIterator(&w);
-    if (itWindow != g_window_list.end())
-        g_window_list.erase(itWindow);
+    w.flags |= WF_DEAD;
+}
+
+void WindowFlushDead()
+{
+    // Remove all windows in g_window_list that have the WF_DEAD flag
+    g_window_list.remove_if([](const std::shared_ptr<WindowBase>& w) -> bool { return w->flags & WF_DEAD; });
```

Comment:
`auto` could make this more readable:
```suggestion
    g_window_list.remove_if([](const auto&& w) -> bool { return w->flags & WF_DEAD; });
```